stl/inc/experimental/random

```diff
@@ -0,0 +1,55 @@
+// random experimental header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _EXPERIMENTAL_RANDOM_
+#define _EXPERIMENTAL_RANDOM_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#include <random>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+namespace experimental {
+    inline namespace fundamentals_v3 {
+
+        inline default_random_engine& _Engine() {
+            thread_local default_random_engine _Eng{random_device{}()};
+            return _Eng;
+        }
+
+        template <class _IntType>
+        _IntType randint(_IntType _Min, _IntType _Max) {
+            static_assert(_Is_any_of_v<remove_cv_t<_IntType>, short, int, long, long long, unsigned short, unsigned int,
+                              unsigned long, unsigned long long>,
+                "template argument does not meet the requirements for IntType.");
+            _STL_ASSERT(_Min <= _Max, "_Min > _Max");
+            return uniform_int_distribution<_IntType>{_Min, _Max}(_Engine());

```

Comment:
This is not correct. You are creating a new uniform_int_distribution with a given range and Call operator() on that temporary. However, the specification speaks from a thread_local instance of a uniform_int_distribution. Now I fail to see how that is possible if there are different calls to randint with different parameters. @BillyONeal might know, at least i remember him talking about rngs.

In any case the initialization of the random number engines/distributions is generally rather costly.So recreating one seems Bad.

Reply:
> In any case the initialization of the random number engines/distributions is generally rather costly.So recreating one seems Bad.

AFAIK distributions aren't costly only engines. And the engine being used is a `static thread_local` (`static` is implied by `thread_local` unless I'm mistaken) instance that's returned by reference.

Reply:
>However, the specification speaks from a thread_local instance of a uniform_int_distribution. 

That's not observable because uniform_int_distribution is stateless.

Reply:
Specifically, in our current implementation, `uniform_int_distribution` is stateless (with respect to the random bits; of course the limits are stored), but the Standard permits random bits to be retained.

Reply:
>but the Standard permits random bits to be retained

It does? Huh, TIL. http://eel.is/c++draft/rand.req.dist#tab:rand.req.dist says there are only requirements on operator() if the same URBG is used successively.

Reply:
> http://eel.is/c++draft/rand.req.dist#tab:rand.req.dist says there are only requirements on operator() if the same URBG is used successively.

That's imposing the uniformity requirement. If you invoke a distribution repeatedly with different engines, then uniformity isn't guaranteed. Indeed it could not be, since I could invoke the distribution repeatedly with the same engine state.

The hint that distributions are allowed to cache some random bits is the specification (and existence) of `reset`, which guarantees that "Subsequent uses of `d` do not depend on values produced by any engine prior to invoking `reset`." (i.e., any cached bits are flushed).